### ISSUE TITLE
CKEditor 5 - Problem creating links

### ISSUE THREAD

**Issue by stpaultim** [backdrop/backdrop-issues#6359]
## Description of the bug

We are in the process of upgrading a Drupal 7 site to Backdrop CMS. We updated to the 1.27 preview release and upgraded to CKEditor 5. 

We were not able to create links until we added the image tool to the toolbar. Once this was added, things worked as expected. However, I think we've decided that this is a bug that needs resolving. 

See: https://github.com/backdrop/backdrop-issues/issues/6359#issuecomment-1884173754

![image](https://github.com/backdrop/backdrop-issues/assets/3144571/fa1ac943-0a0d-48ff-bdf6-cdb11feb4124)

**Comment by stpaultim** [comment]
As I expected, this is not a problem on a fresh site running CKEditor 5. So it might be something specific to our site or it might be something related to the upgrade. 

We'll try to do some more investigations and report our results. But, if anyone has any suggestions, please let us know.

**Comment by stpaultim** [comment]
@indigoxela - Adding the image plugin/icon to toolbar solved the problem. 

This seems like a bug to me, because we were not trying to add a link to an image. There is no obvious connection between these two. 

I'm checking to see if this is an issue on a new site. Maybe it's something about our upgraded configuration?

**Comment by quicksketch** [comment]
@stpaultim Yes this is definitely a bug. @indigoxela probably thinks so too, but by adding the Image button it makes available the missing `ImageUtils` plugin. Looks like we need to wrap a section of logic in an `if` statement to check if images are even supported before trying to add link support to them. Or we could make the `BackdropLink` plugin depend on `ImageUtils` even if it's not always used. The tag filter will still block `<img>` tags if needed and the logic could stay as-is.

**Comment by indigoxela** [comment]
@stpaultim my suggestion to try enabling the image toolbar item wasn't an attempt to claim this isn't a bug. :wink: It is.

It was an attempt to verify that there's something wrong in the BackdropLink vs. ImageUtils logic. Although I didn't see the problem when disabling the image button in toolbar. A bit more digging might be necessary.

**Comment by stpaultim** [comment]
Thanks. That all makes sense?

Let me know if there is something I can test or do to help.

**Comment by quicksketch** [comment]
I filed a PR at https://github.com/backdrop/backdrop/pull/4630 that just loads the missing plugin. Since the entire `ckeditor-dll.js` file is loaded regardless of which plugins are requested, there is no difference in the network file size. And having `ImageUtils` available all the time is a lot easier than several conditional statements to see if it's available. Note loading `ImageUtils` does not mean images are allowed, it just helps detect them if they're present. And it helps in the edge-case of a text format that might allow `<img>` tags even if the `backdropImage` button is not in the toolbar.

**Comment by indigoxela** [comment]
> Let me know if there is something I can test or do to help.

@stpaultim  In fact, there is. I'm still struggling to replicate the console error from your screenshot. What I tried so far: add a linked image in content, then go and remove the image icon from toolbar, then open the node form again and try to add a fresh text link. But that works for me. I do see some nagging in console, but no error and adding or editing links just seems to work fine.

So, while the PR makes sense to me, I can't verify yet, that it fixes the actual problem you're facing.

**Comment by quicksketch** [comment]
I got the error by removing the image button from the toolbar,  then
creating a new piece of content, add some text, select text,  then click
the link button. The process of determining if an image tag is selected
causes the error to occur.

**Comment by indigoxela** [comment]
After multiple flush-cache I had the quick idea to also remove the "Styles" button from the toolbar... Wow. Without that and the heading dropdown I can't type even plain text. :fearful: But that seems unrelated here.

Started fresh (removed editor from text format, added it back after saving), but I get the same result:

Lots of `image-style-missing-dependency...` warnings, but adding links to the form input works just fine.

**Comment by stpaultim** [comment]
@quicksketch - I am not able to recreate the problem using your steps. I originally experienced this problem on a site that had been upgraded from Drupal 7 and had existing content in it after upgrading from CKEditor 4.

I have not yet been able to recreate this problem on a new site that starts with CKEditor 5 as default option.

**Comment by stpaultim** [comment]
@quicksketch or @indigoxela - In trying to recreate this issue on a fresh site, I found a different problem, which may be related or maybe it's a new issue. 

1) Spin up a brand new site with dev branch
2) Remove the "Headings" icon from the CKEditor 5 Toolbar (Basic text editor)
3) Try to create a new page

Note it is impossible to type anything in the body field. 

Do we think this might be related or should I open a new issue for this?

**Comment by herbdool** [comment]
I couldn't recreate the problem either. Must be only under a full moon or something. Even with the PR I see the `image-style-missing-dependency`.

**Comment by stpaultim** [comment]
@herbdool - The problem was originally reported from a site that had been upgraded from Drupal 7 to Backdrop CMS using CKEditor 4 and then CKEditor 4 upgraded to CKEditor 5. So, I expect that there a large number of potential factors in that scenario that might be different from trying to recreate the problem on a fresh Backdrop site.

If I remove the "Image" icon from the toolbar, the problem returns. I have not tried to apply the PR to this site yet to test it. But, I hope to do that tonight.

**Comment by bdalomgir** [comment]
I have applied this [PR](https://github.com/backdrop/backdrop/pull/4630) it fixed the link issue, and I found another issue with the image icon it is unable to add images by clicking the image icon.

![after-clicking-the-image-icon](https://github.com/backdrop/backdrop/assets/13948360/5d3dda43-265a-44b4-9b7a-5b86c26e3753)


**Note:** The image icon is working fine in the fresh backdrop installation.

**Comment by stpaultim** [comment]
@bdalomgir 

Do you know for sure that the image icon worked before you added the PR. Or is it possible that the image icon never worked on this site. 

NOTE: If anyone is not clear @bdalomgir is working on the project with me and this is the site that originally reported this problem.

**Comment by bdalomgir** [comment]
@stpaultim 
1. After adding the PR I can create link in either case by removing the image icon from the toolbar or not.
2. The image icon was broken even if before adding PR also after adding PR, the image icon is working for CKEditor 4.

**Comment by herbdool** [comment]
It's important to try to create the error and the PR with a clean site. You can also test it with an existing site, but only after we've confirmed this all with a clean site.

**Comment by herbdool** [comment]
I can now reliably recreate the error and the fix if I create a new text format before and after.

@quicksketch The PR isn't enough on its own. I had to also enable either "Float images left and right using the data-align attribute" or "Convert image captions to figure and figcaption elements" option.

**Comment by indigoxela** [comment]
> I can now reliably recreate the error and the fix if I create a new text format before and after.

@herbdool can you export the JSON config of such a broken format? Any obvious difference? Or is it a (format) caching problem?

**Comment by herbdool** [comment]
I take it back. I should have taken my own advice and tested with a completely clean site. I still had the contrib ckeditor5 sitting there. Once I removed that completely it worked for me.

**Comment by quicksketch** [comment]
Thanks folks!

> I take it back. I should have taken my own advice and tested with a completely clean site. I still had the contrib ckeditor5 sitting there. Once I removed that completely it worked for me.

Yeah I had that hiccup a few times as well. Since the contrib module "wins" fixes to core don't seem to apply.

I merged https://github.com/backdrop/backdrop/pull/4630 into 1.x for 1.27.0. There are other related issues like https://github.com/backdrop/backdrop-issues/issues/6366 that we'll work on separately.
